src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestTPFatTPFServerWithTranslation.java

```diff
@@ -0,0 +1,223 @@
+				}
+			} else if (i instanceof BGP) {
+				if (unionTranslation == null) {
+					unionTranslation = handleBGP(((BGP) i), fedAccessMgr);
+				} else {
+					unionTranslation = new UnionIteratorForSolMaps(unionTranslation, handleBGP(((BGP) i), fedAccessMgr));
+				}
+			} else {
+				throw new ExecOpExecutionException(i.toString(), this);
+			}
+		}
+		return unionTranslation;
+	}
+	
+	protected Iterator<SolutionMapping> handleGroupPattern(final SPARQLGroupPattern gp, final FederationAccessManager fedAccessMgr) throws ExecOpExecutionException {
+		Iterator<SolutionMapping> groupTranslation = null;
+		for(final SPARQLGraphPattern i : gp.getSubPatterns()) {
+			if (i instanceof TriplePattern) {
+				if (groupTranslation == null) {
+					groupTranslation = handleTriplePattern((TriplePattern) i, fedAccessMgr).iterator();
+				} else {
+					groupTranslation = new JoiningIteratorForSolMaps(getIterableFromIterator(groupTranslation), handleTriplePattern((TriplePattern) i, fedAccessMgr));
+				}
+			} else if (i instanceof SPARQLUnionPattern) {
+				if (groupTranslation == null) {
+					groupTranslation = handleUnionPattern(((SPARQLUnionPattern) i), fedAccessMgr);
+				} else {
+					groupTranslation = new JoiningIteratorForSolMaps(getIterableFromIterator(groupTranslation), getIterableFromIterator(handleUnionPattern(((SPARQLUnionPattern) i), fedAccessMgr)));
+				}
+			} else if (i instanceof SPARQLGroupPattern) {
+				if (groupTranslation == null) {
+					groupTranslation = handleGroupPattern(((SPARQLGroupPattern) i), fedAccessMgr);
+				} else {
+					groupTranslation = new JoiningIteratorForSolMaps(getIterableFromIterator(groupTranslation), getIterableFromIterator(handleGroupPattern(((SPARQLGroupPattern) i), fedAccessMgr)));
+				}
+			} else if (i instanceof BGP) {
+				if (groupTranslation == null) {
+					groupTranslation = handleBGP(((BGP) i), fedAccessMgr);
+				} else {
+					groupTranslation = new JoiningIteratorForSolMaps(getIterableFromIterator(groupTranslation), getIterableFromIterator(handleBGP(((BGP) i), fedAccessMgr)));
+				}
+			} else {
+				throw new ExecOpExecutionException(i.toString(), this);
+			}
+		}
+		return groupTranslation;
+	}
+	
+	protected Iterator<SolutionMapping> handleBGP(final BGP bgp, final FederationAccessManager fedAccessMgr) throws ExecOpExecutionException {
+		Iterator<SolutionMapping> bgpTranslation = null;
+		for(final TriplePattern i : bgp.getTriplePatterns()) {
+			if (bgpTranslation == null) {
+				bgpTranslation = handleTriplePattern((TriplePattern) i, fedAccessMgr).iterator();
+			} else {
+				bgpTranslation = new JoiningIteratorForSolMaps(getIterableFromIterator(bgpTranslation), handleTriplePattern((TriplePattern) i, fedAccessMgr));
+			}
+		}
+		return bgpTranslation;
+	}
+	
+	//Helper class
+	protected static class MaterializingIntermediateResultElementSinkWithTranslation implements ClosableIntermediateResultElementSink {
```

Comment:
Is this class copied from somewhere else? If so, would it be possible to make a sub-class that handles vocabulary mappings and overrides just the send method?